conda_libmamba_solver/shards.py

```diff
@@ -0,0 +1,341 @@
+    """
+
+    def __init__(self, repodata: dict, url: str = ""):
+        """
+        url: for debugging; not used by this class.
+        """
+        all_packages = {
+            "packages": repodata.pop("packages", {}),
+            "packages.conda": repodata.pop("packages.conda", {}),
+        }
+        self.repodata = repodata  # without packages, packages.conda
+        self.url = url
+
+        shards = defaultdict(lambda: {"packages": {}, "packages.conda": {}})
+
+        for package_group in all_packages:
+            for package, record in all_packages[package_group].items():
+                name = record["name"]
+                shards[name][package_group][package] = record
+
+        # defaultdict behavior no longer wanted
+        self.shards: dict[str, Shard] = dict(shards)  # type: ignore
+
+        # used to write out repodata subset
+        self.visited: dict[str, Shard | None] = {}
+
+    def __repr__(self):
+        return self.url.join(super().__repr__().split(maxsplit=1))
+
+    def __contains__(self, package):
+        return package in self.shards
+
+    def fetch_shard(self, package: str, session) -> Shard:
+        """
+        "Fetch" an individual shard.
+
+        Update self.visited with all not-None packages.
+
+        Raise KeyError if package is not in the index.
+        """
+        shard = self.shards[package]
+        self.visited[package] = shard
+        return shard
+
+    def fetch_shards(self, packages: Iterable[str], session) -> dict[str, Shard]:
+        """
+        Fetch multiple shards in one go.
+
+        Update self.visited with all not-None packages.
+        """
+        return {package: self.fetch_shard(package, session) for package in packages}
+
+    def build_repodata(self):
+        """
+        Return monolithic repodata including all visited shards.
+        """
+        repodata = self.repodata.copy()
+        repodata.update({"packages": {}, "packages.conda": {}})
+        for package, shard in self.visited.items():
+            if shard is None:
+                continue  # recorded visited but not available shards
+            for package_group in ("packages", "packages.conda"):
+                repodata[package_group].update(shard[package_group])
+        return repodata
+
+
+FETCHED_THIS_PROCESS = set()
+
+
+class Shards(ShardLike):
+    def __init__(self, shards_index: ShardsIndex, url: str):
+        """
+        Args:
+            shards_index: raw parsed msgpack dict
+            url: URL of repodata_shards.msgpack.zst
+        """
+        self.shards_index = shards_index
+        self.url = url
+
+        self.repodata = {k: v for k, v in self.shards_index.items() if k not in ("shards",)}
+
+        # used to write out repodata subset
+        self.visited: dict[str, Shard | None] = {}
+
+    def __contains__(self, package):
+        return package in self.packages_index
+
+    @property
+    def packages_index(self):
```

Comment:
Return type hint?

Reply:
My editor also knows the return type already?